### PR TITLE
libnetwork: clean up vestigial datastore-related code

### PR DESCRIPTION
**Pull request by corhere**
**- What I did**
I cleaned up code related to datastores and their configuration across libnetwork, pruning dead code and refactoring what remained. Highlights:
  - Refactored package `drvregistry` to no longer be an ad-hoc Dependency Injection container for network and IPAM drivers, and created separate types for network driver registries and IPAM driver registries
    - The `PluginGetter` dependency is now injected directly into the remote drivers
    - Removed the vestiges of datastore dependency injection
  - Removed `DatastoreConfig` from `discoverapi` as dynamic datastore discovery is unreachable
  - Threw out much of the dead persistence code from package `ipam`

**- How I did it**
Prove that code is unreachable. Delete the dead code. Repeat.

**- How to verify it**
CI.

**- Description for the changelog**
N/A

**- A picture of a cute animal (not mandatory but encouraged)**

Cat.

### CODE REVIEW

File: libnetwork/driverapi/driverapi.go [hunk 1]
```diff
@@ -156,12 +156,16 @@ type JoinInfo interface {
...
 }
 
+// Registerer provides a way for network drivers to be dynamically registered.
+type Registerer interface {
+	RegisterDriver(name string, driver Driver, capability Capability) error
+}
+
 // DriverCallback provides a Callback interface for Drivers into LibNetwork
 type DriverCallback interface {
+	Registerer
```
Comment (cpuguy83) [inline libnetwork/driverapi/driverapi.go:166]: lol, this type reminds of https://www.youtube.com/watch?v=6kZBJs527-k
"The Rural Juror"

**Review by cpuguy83 — APPROVED** [review]

File: libnetwork/ipams/remote/remote.go [hunk 2]
```diff
@@ -63,6 +69,7 @@ func Init(cb ipamapi.Callback, l, g interface{}) error {
 	}
 	handleFunc(ipamapi.PluginEndpointType, newPluginHandler)
 	return nil
+
```
Comment (tianon) [inline libnetwork/ipams/remote/remote.go:72]: :eyes:

File: libnetwork/drvregistry/ipams.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/docker/docker/libnetwork/types"
+)
+
+type ipamDriver struct {
+	driver     ipamapi.Ipam
+	capability *ipamapi.Capability
+}
+
+// IPAMs is a registry of IPAM drivers. The zero value is an empty IPAM driver
+// registry ready to use.
```
Comment (neersighted) [inline libnetwork/drvregistry/ipams.go:18]:
```suggestion
// registry, ready to use.
```

**Review by neersighted — CHANGES_REQUESTED** [review]
I noticed one missing comma :laughing:

Otherwise, everything seems well-explained, rational, and well-divided into commits. Awesome work, and it's nice to see that (dead) persistence code just fall away.

File: libnetwork/drvregistry/networks_test.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+		var driverName string
+		reg.WalkDrivers(func(name string, driver driverapi.Driver, capability driverapi.Capability) bool {
+			driverName = name
+			return false
+		})
+
+		assert.Check(t, is.Equal(driverName, mockDriverName))
+	})
+
```
Comment (tianon) [inline libnetwork/drvregistry/networks_test.go:51]: :eyes:

File: libnetwork/drivers/ipvlan/ivmanager/ivmanager.go [hunk 5]
```diff
@@ -11,13 +11,21 @@ const networkType = "ipvlan"
...
+
+// Register registers a new instance of the ipvlan manager driver.
+func Register(r driverapi.Registerer, config map[string]interface{}) error {
 	c := driverapi.Capability{
 		DataScope:         datastore.LocalScope,
 		ConnectivityScope: datastore.GlobalScope,
 	}
-	return dc.RegisterDriver(networkType, &driver{}, c)
+	return r.RegisterDriver(networkType, &driver{}, c)
+
```
Comment (tianon) [inline libnetwork/drivers/ipvlan/ivmanager/ivmanager.go:28]: :eyes:

**Review by tianon — APPROVED** [review]
"Registerer" is a little bit awkward and there are a few trailing newlines at EOfunc, but nothing that I'd personally block this for :sweat_smile:

**Review by vvoland — APPROVED** [review]

**Comment by corhere** [comment]
> "Registerer" is a little bit awkward

Nobody's going to confuse that for a concrete type! [Also:](https://pkg.go.dev/github.com/prometheus/client_golang@v1.14.0/prometheus#Registerer)

**Review by neersighted — APPROVED** [review]

**Pull request merged**